tuf/repository_lib.py

```diff
@@ -1541,6 +1559,168 @@ def _get_hashes_and_length_if_needed(use_length, use_hashes, full_file_path,
+    return True
+  for i in range(s):
+    if a_to_power == n - 1:
+      return True
+    a_to_power = (a_to_power * a_to_power) % n
+  return False
+
+
+
+
+
+def miller_rabin(n, rounds):
+  if n == 1:
+    return false
+  if n == 2 or n == 3:
+    return true
+
+  d = n -1
+  s = 0
+  while d % 2 == 0:
+    d = d >> 1
+    s = s + 1
+
+  for i in range(rounds):
+    a = random.randrange(n)
+    if not miller_rabin_round(a, s, d, n):
+      return False
+  return True
+
+
+
+
+# RSA Accumulator code insprired by https://github.com/ElrondNetwork/elrond-go/blob/v1.0.30/crypto/accumulator/rsa/rsaAcc.go
+def hash_to_prime(data):
+  # TODO: move constant definitions
+  basesMillerRabin = 12
+
+  h = blake2b(str(data).encode('utf-8'))
+  p = int(h.hexdigest(), 16)
+
+  # use Miller-Rabin primality test, if p is not prime, do more rounds of hashing
+  while (not miller_rabin(p, basesMillerRabin)):
+    h = blake2b(str(p).encode('utf-8'))
+    p = int(h.hexdigest(), 16)
+
+  return p
+
+
+
+class acc_contents(object):
+  contents = None
+  name = None
+  proof = None
+
+  def __init__(self, name, contents):
+    # Include the name to ensure the digest differs between elements and cannot be replayed
+    contents["name"] = name
+    self.contents = contents
+    self.name = name
+
+  def set_proof(self, proof):
+    self.proof = proof
+
+
+
+def _build_rsa_acc(fileinfodict):
+  """
+  Create an RSA accululator from the snapshot fileinfo and writes it to individual snapshot files
+
+  Returns the root and leaves
+  """
+
+  # RSA accululator contants
+  # TODO: move constant definitions
+  g = 3
+  # Modulus from https://en.wikipedia.org/wiki/RSA_numbers#RSA-2048
+  # We will want to generate a new one
+  Modulus = "2519590847565789349402718324004839857142928212620403202777713783604366202070759555626401852588078" + \
+		"4406918290641249515082189298559149176184502808489120072844992687392807287776735971418347270261896375014971" + \
+		"8246911650776133798590957000973304597488084284017974291006424586918171951187461215151726546322822168699875" + \
+		"4918242243363725908514186546204357679842338718477444792073993423658482382428119816381501067481045166037730" + \
+		"6056201619676256133844143603833904414952634432190114657544454178424020924616515723350778707749817125772467" + \
+		"962926386356373289912154831438167899885040445364023527381951378636564391212010397122822120720357"
+  m = int(Modulus, 10)
+
+
+  # We will build the accumulator starting with the leaf nodes. Each
+  # leaf contains snapshot information for a single metadata file.
+  leaves = []
+  primes = []
+  acc_exp = 1
+  for name, contents in sorted(fileinfodict.items()):
+    if name.endswith(".json"):
+      name = os.path.splitext(name)[0]
+    cont = acc_contents(name, contents)
+    leaves.append(cont)
+
+    json_contents = securesystemslib.formats.encode_canonical(contents)
+    prime = hash_to_prime(json_contents)
+    primes.append(prime)
+    acc_exp = acc_exp * prime
```

Comment:
This is a hidden O(n^2) algorithm: acc_exp is O(n) in each iteration!

I think there's ways to do this O(n lg n) but need to think about it/dredge them up.

Best trick (O(n)) is to reduce acc_exp mod the order of the group---but that only works if you know the group order (which means knowing the RSA factors---we can think about whether that works in our setting. e.g., if you sign manifests we're kinda implicitly trusting you -- *except* then we lose a little bit of auditability because you can forge proofs if you know the RSA factors. we should discuss).